### ISSUE TITLE
nyx on OpenBSD 6.8 (armv7 architecture build) dies with RuntimeError: dictionary keys changed during iteration

### ISSUE THREAD

**Issue by dgets** [torproject/nyx#36]
I have tried the _package_, _port_, and _pip_ installs of **nyx** version 2.1.0 on the armv7 arch version of OpenBSD 6.8, and have gotten the same error every time.  Upon loading the script, the following stacktrace appears and the user is dumped back to the shell:

```
borgi$ nyx                                                                                                                            
Traceback (most recent call last):                                                                                                    
  File "/usr/local/bin/nyx", line 11, in <module>                                                                                     
    load_entry_point('nyx==2.1.0', 'console_scripts', 'nyx')()                                                                        
  File "/usr/local/lib/python3.8/site-packages/nyx/__init__.py", line 176, in main                                                    
    nyx.starter.main()                                                                                                                
  File "/usr/local/lib/python3.8/site-packages/stem/util/conf.py", line 289, in wrapped                                               
    return func(*args, config = config, **kwargs)                                                                                     
  File "/usr/local/lib/python3.8/site-packages/nyx/starter.py", line 90, in main                                                      
    controller = init_controller(                                                                                                     
  File "/usr/local/lib/python3.8/site-packages/nyx/__init__.py", line 325, in init_controller                                         
    TOR_CONTROLLER = stem.connection.connect(*args, **kwargs)                                                                         
  File "/usr/local/lib/python3.8/site-packages/stem/connection.py", line 291, in connect                                              
    return _connect_auth(control_connection, password, password_prompt, chroot_path, controller)                                      
  File "/usr/local/lib/python3.8/site-packages/stem/connection.py", line 375, in _connect_auth                                        
    return controller(control_socket, is_authenticated = True)                                                                        
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 1057, in __init__                                               
    super(Controller, self).__init__(control_socket, is_authenticated)                                                                
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 585, in __init__                                                
    self._post_authentication()                                                                                                       
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 3902, in _post_authentication                                   
    owning_pid = self.get_conf('__OwningControllerProcess', None)                                                                     
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 2170, in get_conf                                               
    entries = self.get_conf_map(param, default, multiple)                                                                             
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 2273, in get_conf_map                                           
    for key in reply:                                                                                                                 
RuntimeError: dictionary keys changed during iteration                                                                                
borgi$
```

My apologies if this is an error that should be reported somewhere else; if so, please let me know the appropriate group or forum to report it on and I will take it there.  Thank you for your time and assistance on this matter!

**Comment by atagar** [comment]
Hi Damon, what version of Stem are you using? If you 'pip install stem' to upgrade to 1.8 does this problem persist?

**Comment by dgets** [comment]
Hello, Damian, thank you for the quick response!

I was previously using _stem_ 1.7.1.  After upgrading to 1.8 the same problem did, indeed, persist.  I decided for some reason to try running _nyx_ as root, however, and I'm seeing that it's now working, somewhat.  Of course the warnings about running _nyx_ as root not being necessary nor advised are on the display...  Not sure why running as root would change this situation, I know that I tried that at one point before upgrading to _stem_ 1.8 and I was still receiving the error.  Anyway at this point (as root), the bandwidth graph is showing, but there doesn't appear to be any of the usual information for the startup or anything else really in the events log.  The contents that are shown there are copied here, in case it's useful:

> Events (TOR/NYX NOTICE-ERR):
>  18:54:05 [NYX_NOTICE] Tor connection information is unavailable. This is fine, but if you would like to have it please see
>    https://nyx.torproject.org/#no_connections
>  18:53:50 [NYX_NOTICE] Nyx is currently running with root permissions. This isn't a good idea, nor should it be necessary.
>  18:53:50 [NYX_NOTICE] No nyxrc loaded, using defaults. You can customize nyx by placing a configuration file at /root/.nyx/config
>    (see https://nyx.torproject.org/nyxrc.sample for its options).

Guess I'm a bit confounded here.  Anything you might be able to offer would be helpful.  Thanks again!

**Comment by atagar** [comment]
> After upgrading to 1.8 the same problem did, indeed, persist.

Could you please provide a stacktrace (like you did earlier) for when it crashes with Stem 1.8?

**Comment by dgets** [comment]
Here ya go, sir:

```
borgi$ nyx   
Traceback (most recent call last):
  File "/usr/local/bin/nyx", line 11, in <module>
    load_entry_point('nyx==2.1.0', 'console_scripts', 'nyx')()
  File "/usr/local/lib/python3.8/site-packages/nyx/__init__.py", line 176, in main
    nyx.starter.main()
  File "/usr/local/lib/python3.8/site-packages/stem/util/conf.py", line 289, in wrapped
    return func(*args, config = config, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/nyx/starter.py", line 90, in main
    controller = init_controller(
  File "/usr/local/lib/python3.8/site-packages/nyx/__init__.py", line 325, in init_controller
    TOR_CONTROLLER = stem.connection.connect(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/stem/connection.py", line 291, in connect
    return _connect_auth(control_connection, password, password_prompt, chroot_path, controller)
  File "/usr/local/lib/python3.8/site-packages/stem/connection.py", line 375, in _connect_auth
    return controller(control_socket, is_authenticated = True)
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 1057, in __init__
    super(Controller, self).__init__(control_socket, is_authenticated)
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 585, in __init__
    self._post_authentication()
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 3902, in _post_authentication
    owning_pid = self.get_conf('__OwningControllerProcess', None)
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 2170, in get_conf
    entries = self.get_conf_map(param, default, multiple)
  File "/usr/local/lib/python3.8/site-packages/stem/control.py", line 2273, in get_conf_map
    for key in reply:
RuntimeError: dictionary keys changed during iteration
borgi$
```

**Comment by dgets** [comment]
My apologies...  I didn't realize until I started looking at the [identical] stack trace this time around...  Of _course_ it worked previously when I ran it as _root_, because I installed the _stem_ upgrade in _pip_ under _sudo_.  Derp.  I guess it just didn't have the correct options set to populate the event log.

Running it right now as an ordinary user, after I've installed _stem_'s upgrade to 1.8 under that account.  I'm not seeing anything out of the ordinary at all, it seems to be working perfectly.  Thank you for holding my hand through this a bit, I really appreciate it!
